external-crates/move/crates/move-analyzer/src/symbols.rs

```diff
@@ -997,14 +1007,62 @@ pub fn get_symbols(
         file_id_to_lines.insert(id, lines);
     }
 
+    let compiler_flags = resolution_graph.build_options.compiler_flags().clone();
     let build_plan = BuildPlan::create(resolution_graph)?;
     let mut parsed_ast = None;
     let mut typed_ast = None;
     let mut diagnostics = None;
-    build_plan.compile_with_driver(&mut std::io::sink(), |mut compiler| {
-        compiler = compiler.set_vfs_root(overlay_fs_root.clone());
+
+    let mut dependencies = build_plan.compute_dependencies();
+    let compiled_libs = if let Ok(deps_package_paths) = dependencies.make_deps_for_compiler() {
+        // Partition deps_package according whether src is available
+        let src_deps = deps_package_paths
+            .iter()
+            .filter_map(|(p, b)| {
+                if let ModuleFormat::Source = b {
+                    Some(p.clone())
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let src_names = src_deps
+            .iter()
+            .filter_map(|p| p.name.as_ref().map(|(n, _)| *n))
+            .collect::<BTreeSet<_>>();
+
+        let compiled_deps = match pkg_deps.get(pkg_path) {
+            Some(d) => {
+                eprintln!("found pre-compiled libs for {:?}", pkg_path);
+                Some(d.clone())
+            }
+            None => construct_pre_compiled_lib(src_deps, None, compiler_flags)
+                .ok()
+                .and_then(|pprog_and_comments_res| pprog_and_comments_res.ok())
+                .map(|lib| {
+                    eprintln!("created pre-compiled libs for {:?}", pkg_path);
+                    let res = Arc::new(lib);
+                    pkg_deps.insert(pkg_path.to_path_buf(), res.clone());
+                    res
+                }),
+        };
+        if compiled_deps.is_some() {
+            // if successful, remove only source deps but keep bytecode deps as they
+            // were not used to construct pre-compiled lib in the first place
+            dependencies.remove_deps(src_names);
+        }
+        compiled_deps
+    } else {
+        None
+    };
+
+    build_plan.compile_with_driver_and_deps(dependencies, &mut std::io::sink(), |compiler| {
         // extract expansion AST
-        let (files, compilation_result) = compiler.run::<PASS_PARSER>()?;
+        let (files, compilation_result) = compiler
+            .set_vfs_root(overlay_fs_root.clone())
+            .set_pre_compiled_lib_opt(compiled_libs.clone())
```

Comment:
If we change thos ti `compiled_libs.as_ref()`, can we avoid much of the work to change everything to `Arc`s?